google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageImpl.java

```diff
@@ -936,4 +983,45 @@ private static String fmtMethodName(String name, Class<?>... args) {
         + Arrays.stream(args).map(Class::getName).collect(Collectors.joining(", "))
         + ")";
   }
+
+  private ReadObjectRequest getReadObjectRequest(
+      BlobId blob, Map<StorageRpc.Option, ?> optionsMap) {
+    Object object = codecs.blobId().encode(blob);
+
+    ReadObjectRequest.Builder builder =
+        ReadObjectRequest.newBuilder().setBucket(object.getBucket()).setObject(object.getName());
+
+    long generation = object.getGeneration();
+    if (generation > 0) {
+      builder.setGeneration(generation);
+    }
+    ifNonNull(
+        (Long) optionsMap.get(StorageRpc.Option.IF_METAGENERATION_MATCH),
+        builder::setIfMetagenerationMatch);
+    ifNonNull(
+        (Long) optionsMap.get(StorageRpc.Option.IF_METAGENERATION_NOT_MATCH),
+        builder::setIfMetagenerationNotMatch);
+    ifNonNull(
+        (Long) optionsMap.get(StorageRpc.Option.IF_GENERATION_MATCH),
+        builder::setIfGenerationMatch);
+    ifNonNull(
+        (Long) optionsMap.get(StorageRpc.Option.IF_GENERATION_NOT_MATCH),
+        builder::setIfGenerationNotMatch);
+    String userProject = (String) optionsMap.get(StorageRpc.Option.USER_PROJECT);
+    if (userProject != null) {
+      // TODO: user project
+      // ReadObjectRequest does not have a user_project field
```

Comment:
IIUC, we need to use gRPC Metadata for this now per cl/448634966

I see: https://g3doc.corp.google.com/net/grpc/g3doc/java/life_of_a_stream_java.md?cl=head#clientinterceptor 

Looks like we need to use a client side interceptor but couldn't figure out where to attach it. Maybe you'll have better luck.